File: apps/frontend/src/contracts/schemas.ts
import { z } from "zod";

export const JobStatusSchema = z.enum([
  "queued",
  "downloading",
  "downloaded",
  "extracting_audio",
  "generating_thumbnail",
  "transcribing",
  "summarizing",
  "generating_tags",
  "completed",
  "failed",
]);

export type JobStatus = z.infer<typeof JobStatusSchema>;

export const ChapterSchema = z.object({
  start: z.string().min(1),
  title: z.string().min(1),
});

export type Chapter = z.infer<typeof ChapterSchema>;

export const CreateVideoRequestSchema = z.object({
  url: z.string().min(1),
});

export type CreateVideoRequest = z.infer<typeof CreateVideoRequestSchema>;

export const VideoJobResponseSchema = z.object({
  id: z.string(),
  url: z.string(),
  title: z.string().nullable().optional(),
  status: JobStatusSchema.default("queued"),
  current_step: z.string().default("queued"),
  video_path: z.string().nullable().optional(),
  audio_path: z.string().nullable().optional(),
  thumbnail_path: z.string().nullable().optional(),
  transcript: z.string().nullable().optional(),
  summary: z.string().nullable().optional(),
  tags: z.array(z.string()).default([]),
  chapters: z.array(ChapterSchema).default([]),
  error: z.string().nullable().optional(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

export type VideoJobResponse = z.infer<typeof VideoJobResponseSchema>;

export const WorkerMessageSchema = z.object({
  job_id: z.string(),
  video_id: z.string(),
  attempt: z.number().int().default(1),
  url: z.string().optional(),
});

export type WorkerMessage = z.infer<typeof WorkerMessageSchema>;

File: packages/contracts/src/tubedigest/contracts/schemas.py
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field
from typing import Optional


class JobStatus(str, Enum):
    queued = "queued"
    downloading = "downloading"
    downloaded = "downloaded"
    extracting_audio = "extracting_audio"
    generating_thumbnail = "generating_thumbnail"
    transcribing = "transcribing"
    summarizing = "summarizing"
    generating_tags = "generating_tags"
    completed = "completed"
    failed = "failed"


class Chapter(BaseModel):
    start: str = Field(min_length=1)
    title: str = Field(min_length=1)


class CreateVideoRequest(BaseModel):
    url: str = Field(min_length=1)


class VideoJobResponse(BaseModel):
    id: str
    url: str
    title: Optional[str] = None
    status: JobStatus = JobStatus.queued
    current_step: str = "queued"
    video_path: Optional[str] = None
    audio_path: Optional[str] = None
    thumbnail_path: Optional[str] = None
    transcript: Optional[str] = None
    summary: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    chapters: list[Chapter] = Field(default_factory=list)
    error: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class WorkerMessage(BaseModel):
    job_id: str
    video_id: str
    attempt: int = 1
    url: Optional[str] = None
